source/agora/consensus/validation/Block.d

```diff
@@ -112,6 +119,281 @@ public string isInvalidReason (const ref Block block, in ulong prev_height,
+        buildMerkleTree(block);
+        assert(block.isGenesisBlockInvalid());
+
+        // Txs sorting check
+        block.txs.reverse;
+        buildMerkleTree(block);
+        assert(!block.isGenesisBlockInvalid());
+
+        block.txs.reverse;
+        buildMerkleTree(block);
+        assert(block.isGenesisBlockInvalid());
+
+        // Txs length out of bounds check
+        block.txs ~= makeNewTx();
+        block.txs.sort;
+        buildMerkleTree(block);
+        assert(block.txs.length == Block.TxsInBlock + 1);
+        assert(!block.isGenesisBlockInvalid());
+
+        block = GenesisBlock.serializeFull.deserializeFull!Block;
+
+        // Txs type check
+        block.txs[0].type = cast(TxType)2;
+        buildMerkleTree(block);
+        assert(!block.isGenesisBlockInvalid());
+
+        block.txs[0].type = TxType.Payment;
+        buildMerkleTree(block);
+        assert(block.isGenesisBlockInvalid());
+
+        block.txs[0].type = TxType.Freeze;
+        buildMerkleTree(block);
+        assert(block.isGenesisBlockInvalid());
+
+        // Input empty check
+        block.txs[0].inputs ~= Input.init;
+        buildMerkleTree(block);
+        assert(!block.isGenesisBlockInvalid());
+
+        block.txs[0] = tx;
+        buildMerkleTree(block);
+        assert(block.isGenesisBlockInvalid());
+
+        // Output not empty check
+        Output[] outputs;
+        block.txs[0].outputs = outputs;
+        buildMerkleTree(block);
+        assert(!block.isGenesisBlockInvalid());
+
+        // disallow 0 amount
+        Output zeroOutput =
+            Output(Amount.invalid(0), WK.Keys[0].address);
+        block.txs[0].outputs ~= zeroOutput;
+        buildMerkleTree(block);
+        assert(!block.isGenesisBlockInvalid());
+    }
+
+    block = GenesisBlock.serializeFull.deserializeFull!Block;
+
+    // enrollments validation test
+    Enrollment[] enrolls;
+    enrolls ~= Enrollment.init;
+    block.header.enrollments = enrolls;
+    assert(!block.isGenesisBlockInvalid());
+
+    block.header.enrollments.length = 0;
+    assert(block.isGenesisBlockInvalid());
+
+    block = GenesisBlock.serializeFull.deserializeFull!Block;
+
+    // modify the last hex byte of the merkle root
+    block.header.merkle_root[][$ - 1]++;
+    assert(!block.isGenesisBlockInvalid());
+
+    // now restore it back to what it was
+    block.header.merkle_root[][$ - 1]--;
+    assert(block.isGenesisBlockInvalid());
+    const last_root = block.header.merkle_root;
+
+    // the previous merkle root should not match the new txs
+    block.txs ~= newCoinbaseTX(keypair.random().address, Amount(100));
+    block.header.merkle_root = last_root;
+    assert(!block.isGenesisBlockInvalid());
+}
+
+/// Ditto but returns `bool`, only usable in unittests
+/// Only the genesis block Validation
+version (unittest)
+public bool isGenesisBlockInvalid (const ref Block genesis_block)
+    nothrow @safe
+{
+    return isGenesisBlockInvalidReason(genesis_block) is null;
```

Comment:
This should either be `!is null` or renamed to `isGensisBlockValid`. Look at your call below:

```D
assert(GenesisBlock.isGenesisBlockInvalid());
```

It's asserting that the default genesis block is invalid.

Reply:
The meaning of the test function was the opposite.
It's a good point.

I changed its name to `isGenesisBlockValid`
I think there is a typo in the name you recommended :)